### PR TITLE
[ECHOJS-148] serialize methods  and arguments

### PR DESCRIPTION
**Pull request by alesPoz**

### CODE REVIEW

File: src/echo/ws-api/wallet-api.js [hunk 1]
```diff
@@ -382,55 +349,45 @@ class WalletAPI {
 	 * @returns {Promise<[string, string][]>}
 	 */
 	listAccounts(accountName, limit = API_CONFIG.LIST_ACCOUNTS_DEFAULT_LIMIT) {
-		if (!isString(accountName)) throw new Error('account name should be a string');
-		if (!isUInt32(limit) || limit > API_CONFIG.LIST_ACCOUNTS_MAX_LIMIT) {
-			throw new Error(`Limit should be a integer and must not exceed ${API_CONFIG.LOOKUP_ACCOUNTS_MAX_LIMIT}`);
+		if (!limit > API_CONFIG.LIST_ACCOUNTS_MAX_LIMIT) {
```
Comment (nicklatkovich) [inline src/echo/ws-api/wallet-api.js]: `!limit` is always equals to `false` or `true`. So this field will be valid in any way.
Remove `!` here and below.

File: src/serializers/basic/VariantObject.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import ISerializer from '../ISerializer';
+
+/** @typedef {SerializerInput<array.<any>>} TInput */
```
Comment (nicklatkovich) [inline src/serializers/basic/VariantObject.js]:
```js
 /** @typedef {[string, any]} TInput */
```

File: src/serializers/basic/VariantObject.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+/** @typedef {array.<any>} TOutput */
```
Comment (nicklatkovich) [inline src/serializers/basic/VariantObject.js]:
```js
/** @typedef {[string, any]} TOutput */
```

File: src/serializers/basic/VariantObject.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+/** @augments {ISerializer<TInput, TOutput>} */
+export default class VariantObjectSerializer extends ISerializer {
+
+	/**
+	 * @param {TInput<T>} value
```
Comment (nicklatkovich) [inline src/serializers/basic/VariantObject.js]: Genesis is not needed here
```js
* @param {TInput} value
```

File: src/serializers/basic/VariantObject.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+	 * @returns {TOutput<T>}
+	 */
+	toRaw(value) {
+		if (!Array.isArray(value)) throw new Error('value is not an array');
+		if (value.length !== 2) throw new Error('VariantObject should have 2 elements');
+		return value;
```
Comment (nicklatkovich) [inline src/serializers/basic/VariantObject.js:18]: validate, that the first element is string

File: src/serializers/basic/VariantObject.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+/** @typedef {SerializerInput<array.<any>>} TInput */
+/** @typedef {array.<any>} TOutput */
+
+/** @augments {ISerializer<TInput, TOutput>} */
+export default class VariantObjectSerializer extends ISerializer {
+
+	/**
+	 * @param {TInput<T>} value
+	 * @returns {TOutput<T>}
```
Comment (nicklatkovich) [inline src/serializers/basic/VariantObject.js]: rm genesis

File: src/serializers/basic/index.js [hunk 7]
```diff
@@ -14,5 +15,6 @@ export const bytes = (bytesCount) => new BytesSerializer(bytesCount);
 
 export const string = new StringSerializer();
 export const timePointSec = new TimePointSecSerializer();
+export const variantObject = new VariantObjectSerializer();
 
 export { integers, BoolSerializer, BytesSerializer, TimePointSecSerializer };
```
Comment (nicklatkovich) [inline src/serializers/basic/index.js]: export `VariantObjectSerializer`

File: src/serializers/index.js [hunk 8]
```diff
@@ -8,6 +8,9 @@ import opWrapper from './protocol/opWrapper';
 
 export { default as transaction, signedTransactionSerializer as signedTransaction } from './transaction';
 
+export { approvalDelta } from './wallet';
+export { accountListing } from './protocol/account/index';
```
Comment (nicklatkovich) [inline src/serializers/index.js]: export `accountListing` in `protocol/index` (like `price` of asset)
rm from here

File: src/serializers/index.js [hunk 9]
```diff
@@ -8,6 +8,9 @@ import opWrapper from './protocol/opWrapper';
 
 export { default as transaction, signedTransactionSerializer as signedTransaction } from './transaction';
 
+export { approvalDelta } from './wallet';
```
Comment (nicklatkovich) [inline src/serializers/index.js]: make wallet module (like `protocol`)

File: src/serializers/index.js (hunk 8 above)
Comment (nicklatkovich) [reply src/serializers/index.js]: `AccountListingSerializer` and `ACCOUNT_LISTING` too

File: types/serializers/basic/VariantObject.d.ts [hunk 10]
```diff
@@ -0,0 +1,9 @@
+import ISerializer, { SerializerInput } from "../ISerializer";
+
+type TInput = Array<any>;
```
Comment (nicklatkovich) [inline types/serializers/basic/VariantObject.d.ts]:
```ts
type TInput = [string, any];
```

File: types/serializers/basic/VariantObject.d.ts [hunk 11]
```diff
@@ -0,0 +1,9 @@
...
+type TOutput = Array<any>;
```
Comment (nicklatkovich) [inline types/serializers/basic/VariantObject.d.ts]:
```ts
type TOutput = [string, any];
```

File: src/serializers/basic/VariantObject.js [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+	 */
+	toRaw(value) {
+		if (!Array.isArray(value)) throw new Error('value is not an array');
+		if (value.length !== 2) throw new Error('VariantObject should have 2 elements');
+		return value;
+	}
+
+	appendToByteBuffer() {
```
Comment (nicklatkovich) [inline src/serializers/basic/VariantObject.js:21]: You need to implement `readFromBuffer` abstract method like this one

Comment (nicklatkovich) [reply src/serializers/basic/VariantObject.js:21]: In `PrivateKeySerializer` too

**Review by nicklatkovich — CHANGES_REQUESTED** [review]

File: src/serializers/index.js (hunk 9 above)
Comment (alesPoz) [reply src/serializers/index.js]: Sorry, i didn't catch your idea

File: types/serializers/wallet.d.ts [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+declare const approvalDelta : StructSerializer<{
+    active_approvals_to_add: VectorSerializer<StringSerializer>,
+    active_approvals_to_remove: VectorSerializer<StringSerializer>,
+    owner_approvals_to_add: VectorSerializer<StringSerializer>,
+    owner_approvals_to_remove: VectorSerializer<StringSerializer>,
+    key_approvals_to_add: VectorSerializer<StringSerializer>,
+    key_approvals_to_remove: VectorSerializer<StringSerializer>,
+}>;
+
+export default approvalDelta;
```
Comment (nicklatkovich) [inline types/serializers/wallet.d.ts]: Remove `export default` and use `export` (your declaration files should duplicate structure of JS-files)

File: src/serializers/protocol/account/account_listing.js [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+export default class AccountListingSerializer extends UInt8Serializer {
+
+	/**
+	 * @param {TInput<T>} value
+	 * @returns {TOutput<T>}
+	 */
+	toRaw(value) {
+		const result = super.toRaw(value);
+		if (result > 4) throw new Error('invalid account listing');
+	}
```
Comment (nicklatkovich) [inline src/serializers/protocol/account/account_listing.js:28]: Function does not returns value

File: types/serializers/protocol/account/account_listing.d.ts [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+
+export enum ACCOUNT_LISTING {
+    NO_LISTNING =  0x0,
+    WHITE_LISTED = 0x1,
+    BLACK_LISTED = 0x2,
+    WHITE_AND_BLACK_LISTED = 0x3,
+}
+
+export default class AccountListingSerializer extends UInt8Serializer {
+    toRaw(value: number): number;
```
Comment (nicklatkovich) [inline types/serializers/protocol/account/account_listing.d.ts]: `number` >> `ACCOUNT_LISTING`

File: src/serializers/protocol/account/account_listing.js [hunk 16]
```diff
@@ -0,0 +1,25 @@
+import { UInt8Serializer } from '../../basic/integers';
+
```
Comment (nicklatkovich) [inline src/serializers/protocol/account/account_listing.js:2]: Create jsdoc-enum type
```ts
/**
 * @typedef {Object} ACCOUNT_LISTING_t
 * @property {0x0} NO_LISTNING
 * @property {0x1} WHITE_LISTED
 * @property {0x2} BLACK_LISTED
 * @property {0x3} WHITE_AND_BLACK_LISTED
 */
```

Comment (nicklatkovich) [reply src/serializers/protocol/account/account_listing.js:2]: And add this type as type of const ACCOUNT_LISTING
```ts
/** @type {ACCOUNT_LISTING_t} */
 export const ACCOUNT_LISTING = {
// ...
```

Comment (nicklatkovich) [reply src/serializers/protocol/account/account_listing.js:2]: And fix JSDoc of `toRaw` method to
```ts
/**
 * @template {ACCOUNT_LISTING_t[keyof ACCOUNT_LISTING_t]} T
 * @param {T} value
 * @returns {T}
 */```

File: src/serializers/index.js [hunk 17]
```diff
@@ -5,11 +5,12 @@ import * as plugins from './plugins';
 import * as protocol from './protocol';
 import OperationSerializer from './operation';
 import opWrapper from './protocol/opWrapper';
+import * as approvalDelta from './wallet';
```
Comment (nicklatkovich) [inline src/serializers/index.js]: Invalid input format. To use approvalDelta i will need to use something like:
```ts
import { serializers } from "echojs-lib"
serializers.approvalDelta.approvalDelta
```
Change name of this import to wallet to use this serializer like:
```ts
import { serializers } from "echojs-lib"
serializers.wallet.approvalDelta
```

Comment (nicklatkovich) [reply src/serializers/index.js]: And add changes of this files to declaration files

File: src/serializers/chain/index.js [hunk 18]
```diff
@@ -1,5 +1,6 @@
 import * as ids from './id';
```
Comment (nicklatkovich) [inline src/serializers/chain/index.js:1]: Add changes of this files to declaration-files

File: src/serializers/basic/index.js [hunk 19]
```diff
@@ -3,6 +3,7 @@ import BoolSerializer from './Bool';
 import BytesSerializer from './Bytes';
 import StringSerializer from './StringSerializer';
 import TimePointSecSerializer from './TimePointSec';
+import VariantObjectSerializer from './VariantObject';
```
Comment (nicklatkovich) [inline src/serializers/basic/index.js:6]: Add in delcaration-files

**Review by nicklatkovich — CHANGES_REQUESTED** [review]

File: src/serializers/index.js [hunk 20]
```diff
@@ -5,11 +5,12 @@ import * as plugins from './plugins';
...
+import * as wallet from './wallet';
 
 export { default as transaction, signedTransactionSerializer as signedTransaction } from './transaction';
 
 export const operation = new OperationSerializer();
 
 opWrapper.init(operation);
 
-export { basic, collections, chain, OperationSerializer, plugins, protocol };
+export { basic, collections, chain, OperationSerializer, plugins, protocol, opWrapper, wallet };
```
Comment (nicklatkovich) [inline src/serializers/index.js:16]: exporting of wallet is not provided in declaration files

File: src/serializers/protocol/index.js [hunk 21]
```diff
@@ -63,6 +63,8 @@ import {
 
 import VoteIdSerializer from './VoteId';
 
+export { AccountListingSerializer, accountListing, ACCOUNT_LISTING } from './account';
```
Comment (nicklatkovich) [inline src/serializers/protocol/index.js:66]: this code is not provided in declaration files

File: src/echo/ws-api/wallet-api.js [hunk 22]
```diff
@@ -954,70 +853,63 @@ class WalletAPI {
...
-		if (!isBoolean(shouldDoBroadcastToNetwork)) return Promise.reject(new Error('Broadcast should be a boolean'));
-
-		return this.wsRpc.call([0, 'approve_proposal',
-			[feePayingAccountId, proposalId, delta, shouldDoBroadcastToNetwork],
-		]);
+	approveProposal(feePayingAccountId, idOfProposal, delta, shouldDoBroadcastToNetwork) {
+		return this.wsRpc.call([0, 'approve_proposal', [
+			accountId.toRaw(feePayingAccountId),
+			proposalId.toRaw(idOfProposal),
+			wallet.toRaw(delta),
```
Comment (nicklatkovich) [inline src/echo/ws-api/wallet-api.js]: `wallet` is collection of wallet-serializators. It has no method `toRaw`. Use `wallet.approvalDelta.toRaw`.

**Review by nicklatkovich — CHANGES_REQUESTED** [review]

**Review by nicklatkovich — APPROVED** [review]

**Pull request merged**